sources/engine/Stride.Input/Windows/RawInput/RawInput.cs

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) Stride contributors (https://stride3d.net)
+// Distributed under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Stride.Input.RawInput
+{
+    internal static class RawInput
+    {
+        [StructLayout(LayoutKind.Sequential)]
+        private struct RawInputDevice
+        {
+            public ushort UsagePage;
+
+            public ushort Usage;
+
+            public uint Flags;
+
+            public IntPtr Target;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        public struct RawInputHeader
+        {
+            public int dwType;
+            public int dwSize;
+            public IntPtr hDevice;
+            public IntPtr wParam;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        public struct RawMouse
+        {
+            [StructLayout(LayoutKind.Explicit)]
+            public struct RawMouseButtonsData
+            {
+                [FieldOffset(0)]
+                public int Buttons;
+
+                [FieldOffset(0)]
+                public short ButtonFlags;
+
+                [FieldOffset(2)]
+                public short ButtonData;
+            }
+
+            public ushort Flags;
+            public RawMouseButtonsData ButtonsData;
+            public int RawButtons;
+            public int LastX;
+            public int LastY;
+            public int ExtraInformation;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        public struct RawKeyboard
+        {
+            public ushort MakeCode;
+            public ushort Flags;
+            public ushort Reserved;
+            public ushort VKey;
+            public uint Message;
+            public uint ExtraInformation;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        public struct RawHid
+        {
+            public uint dwSizeHid;
+            public uint dwCount;
+            public int bRawData;
+        }
+
+        [StructLayout(LayoutKind.Explicit)]
+        public struct RawInputDataInner
+        {
+            [FieldOffset(0)]
+            public RawMouse Mouse;
+
+            [FieldOffset(0)]
+            public RawKeyboard Keyboard;
+
+            [FieldOffset(0)]
+            public RawHid Hid;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        public struct RawInputData
+        {
+            public RawInputHeader header;
+            public RawInputDataInner data;
+        }
+
+        public static unsafe byte[] GetHIDRawData(ref RawHid rawHid)
+        {
+            var byteCount = rawHid.dwCount * rawHid.dwSizeHid;
+            var result = new byte[byteCount];
```

Comment:
This would result in an allocation (although a small one) potentially every frame, or at least everytime a message is received from a HID. An alternative would be to move the ownership of the buffer outside the function.

This is not very important though. Maybe a good point to keep in mind for a further optimization pass.

Reply:
Yes it would be good to fix.
Maybe it can be stackalloc on the caller side? (assuming it can't be too large to avoid stackoverflow)

Reply:
IMO this can be merged for now. This allocation only happens when a HID device event occurs which is not the case with the current usage. Even in the future it will be rare that it will be used because the HID devices would either be accessed through their own API or through other input libraries such as directinput/xinput.

Reply:
Sorry, I thought it was happening all the time.
Totally fine then :)